File: anvil/src/test/java/trikita/anvil/SyntaxTest.java
package trikita.anvil;

import org.junit.Test;
import static org.junit.Assert.*;

import static trikita.anvil.BaseDSL.*;

public class SyntaxTest extends Utils {

    private boolean showNestedLayout = true;
    private int numberOfChildViews = 5;

    @Test
    public void testSyntaxWithLambdas() {
        // With Java 8 or Kotlin the code below looks much better.
        // Here I use old "lambda"-like syntax to avoid dependency
        // on RetroLambda or Kotlin in Anvil core
        Anvil.mount(container, new Anvil.Renderable() {
            public void view() {
                v(MockLayout.class, new Anvil.Renderable() {
                    public void view() {
                        attr("tag", 1);
                        v(MockView.class, new Anvil.Renderable() {
                            public void view() {
                                attr("tag", 2);
                            }
                        });
                        if (showNestedLayout) {
                            v(MockLayout.class, new Anvil.Renderable() {
                                public void view() {
                                    for (int i = 0; i < numberOfChildViews; i++) {
                                        final int index = i;
                                        v(MockView.class, new Anvil.Renderable() {
                                            public void view() {
                                                attr("tag", index);
                                            }
                                        });
                                    }
                                }
                            });
                        }
                    }
                });
            }
        });
        MockLayout layout = (MockLayout) container.getChildAt(0);
        assertEquals(1, layout.getTag());
        MockView header = (MockView) layout.getChildAt(0);
        assertEquals(2, header.getTag());
        MockLayout content = (MockLayout) layout.getChildAt(1);
        for (int i = 0; i < numberOfChildViews; i++) {
            MockView v = (MockView) content.getChildAt(i);
            assertEquals(i, v.getTag());
        }
    }

    @Test
    public void testSyntaxWithoutLambdas() {
        Anvil.mount(container, new Anvil.Renderable() {
            public void view() {
                o (v(MockLayout.class),
                        attr("tag", 1),
                        o (v(MockView.class),
                                attr("tag", 2)),
                        o (v(MockLayout.class),
                                showNestedLayout ? childViews(numberOfChildViews) : null));
            }
        });
        MockLayout layout = (MockLayout) container.getChildAt(0);
        assertEquals(1, layout.getTag());
        MockView header = (MockView) layout.getChildAt(0);
        assertEquals(2, header.getTag());
        MockLayout content = (MockLayout) layout.getChildAt(1);
        for (int i = 0; i < numberOfChildViews; i++) {
            MockView v = (MockView) content.getChildAt(i);
            assertEquals(i, v.getTag());
        }
    }

    private Void childViews(int count) {
        for (int i = 0; i < numberOfChildViews; i++) {
            o (v(MockView.class), attr("tag", i));
        }
        return null;
    }
}

File: anvil/src/test/kotlin/dev/inkremental/SyntaxTest.kt
package dev.inkremental

import dev.inkremental.dsl.android.view.ViewScope
import org.junit.Assert
import kotlin.test.Test

class SyntaxTest : Utils() {
    private val showNestedLayout = true
    private val numberOfChildViews = 5

    @Test
    fun testSyntax() {
        Inkremental.mount(container, Renderable {
            v<MockLayout> {
                attr("tag", 1)
                v<MockView> { attr("tag", 2) }
                if (showNestedLayout) {
                    v<MockLayout> {
                        for (i in 0 until numberOfChildViews) {
                            v<MockView> { attr("tag", i) }
                        }
                    }
                }
            }
        })
        val layout = container.getChildAt(0) as MockLayout
        Assert.assertEquals(1, layout.tag)
        val header = layout.getChildAt(0) as MockView
        Assert.assertEquals(2, header.tag)
        val content = layout.getChildAt(1) as MockLayout
        for (i in 0 until numberOfChildViews) {
            val v = content.getChildAt(i) as MockView
            Assert.assertEquals(i, v.tag)
        }
    }

    private fun childViews(count: Int) {
        repeat(numberOfChildViews) {
            v<MockView, ViewScope>(ViewScope) { tag(it) }
        }
    }
}
